c++/src/capnp/compat/json.c++

```diff
@@ -882,4 +886,430 @@ void JsonCodec::addFieldHandlerImpl(StructSchema::Field field, Type type, Handle
+
+static constexpr uint64_t JSON_NAME_ANNOTATION_ID = 0xfa5b1fd61c2e7c3dull;
+static constexpr uint64_t JSON_FLATTEN_ANNOTATION_ID = 0x82d3e852af0336bfull;
+static constexpr uint64_t JSON_DISCRIMINATOR_ANNOTATION_ID = 0xcfa794e8d19a0162ull;
+
+class JsonCodec::AnnotatedHandler final: public JsonCodec::Handler<DynamicStruct> {
+public:
+  AnnotatedHandler(JsonCodec& codec, StructSchema schema, kj::Maybe<kj::StringPtr> discriminator,
+                   kj::Vector<Schema>& dependencies)
+      : schema(schema) {
+    auto schemaProto = schema.getProto();
+    auto typeName = schemaProto.getDisplayName();
+
+    if (discriminator == nullptr) {
+      // There are two cases of unions:
+      // * Named unions, which are special cases of named groups. In this case, the union may be
+      //   annotated by annotating the field. In this case, we receive a non-null `discriminator`
+      //   as a constructor parameter, and schemaProto.getAnnotations() must be empty because
+      //   it's not possible to annotate a group's type (becaues the type is anonymous).
+      // * Unnamed unions, of which there can only be one in any particular scope. In this case,
+      //   the parent struct type itself is annotated.
+      // So if we received `null` as the constructor parameter, check for annotations on the struct
+      // type.
+      for (auto anno: schemaProto.getAnnotations()) {
+        switch (anno.getId()) {
+          case JSON_DISCRIMINATOR_ANNOTATION_ID:
+            discriminator = anno.getValue().getText();
+            break;
+        }
+      }
+    }
+
+    KJ_IF_MAYBE(d, discriminator) {
+      unionTagName = discriminator;
+      fieldsByName.insert(std::make_pair(*d, FieldNameInfo {
+        FieldNameInfo::UNION_TAG, 0, 0, nullptr
+      }));
+    }
+
+    discriminantOffset = schemaProto.getStruct().getDiscriminantOffset();
+
+    fields = KJ_MAP(field, schema.getFields()) {
+      auto fieldProto = field.getProto();
+      auto type = field.getType();
+      auto fieldName = fieldProto.getName();
+
+      FieldNameInfo nameInfo;
+      nameInfo.index = field.getIndex();
+      nameInfo.type = FieldNameInfo::NORMAL;
+      nameInfo.prefixLength = 0;
+
+      FieldInfo info;
+      info.name = fieldName;
+
+      kj::Maybe<kj::StringPtr> subDiscriminator;
+      bool flattened = false;
+      for (auto anno: field.getProto().getAnnotations()) {
+        switch (anno.getId()) {
+          case JSON_NAME_ANNOTATION_ID:
+            info.name = anno.getValue().getText();
+            break;
+          case JSON_FLATTEN_ANNOTATION_ID:
+            KJ_REQUIRE(type.isStruct(), "only struct types can be flattened", fieldName, typeName);
+            flattened = true;
+            info.prefix = anno.getValue().getStruct().getAs<JsonFlattenFlags>().getPrefix();
+            break;
+          case JSON_DISCRIMINATOR_ANNOTATION_ID:
+            KJ_REQUIRE(fieldProto.isGroup(), "only unions can have discriminator");
```

Comment:
Does isGroup() being true guarantee that this field is a union, as opposed to a group in a struct? I'm not super familiar with the schema code.

Reply:
No but distinguishing a union from a non-union group is not simple so whatever.